### ISSUE TITLE
Obtain or create code for generating CPS weights file

### ISSUE THREAD

**Issue by MattHJensen** [open-source-economics/taxdata#125]
In #108, @andersonfrailey said: 

> This is because the weights file we have for the CPS only goes to 2026. John gave us the code he used to create the file, but not the code to create the weights file so I can't generate a new one that goes to 2027 on my own.

> I will also email John to see if he has a new weights file available to goes to 2027.

Given that we need to be able to generate the weights file ourselves, shouldn't we be asking John for the code to create the weights file or guidance on how to rewrite that rather than a new weights file? 

I am putting this in its own issue to reflect its high priority and intrinsic importance. This seems exceptionally important to me since it is blocking several other enhancements at this point, and we also can't truly vouch for the CPS file (or even say that it is open source) if we can't build it from scratch. 

Please let me know if I am missing anything or can do anything to assist in obtaining this file.

@Amy-Xu @hdoupe @martinholmer @andersonfrailey @GoFroggyRun

**Comment by andersonfrailey** [comment]
I agree with @MattHJensen. When I get in touch with John I will also ask him for the code he's using.

**Comment by Amy-Xu** [comment]
I agree we definitely need to ask John for the code. Given the average turnaround time potentially involved with that route, would it be possible to use 2026 weight for 2027 as a short-term transitional solution? I assume most extrapolations are done through a linear blow-up with minimum adjustment on weights and therefore adjacent years far out in the future would carry quite similar weights. Certainly I could be mistaken.

@andersonfrailey @MattHJensen

**Comment by martinholmer** [comment]
@andersonfrailey,  When you get to the place where you can generate a `cps_weights.csv` file, be sure to correct this bug:
```
$ awk -F, 'NR==1' cps_weights.csv
SEQUENCE,WT2015,WT2016,WT2017,WT2018,WT2019,WT2020,WT2021,WT2022,WT2023,WT2024,WT2025,WT2026,WT2027

$ awk -F, 'NR==1' puf_weights.csv
WT2009,WT2010,WT2011,WT2012,WT2013,WT2014,WT2015,WT2016,WT2017,WT2018,WT2019,WT2020,WT2021,WT2022,WT2023,WT2024,WT2025,WT2026,WT2027
```
Notice the bogus variable name for the first column in the `cps_weights.csv` file.
It should be `WT2014` not `SEQUENCE`.

This bug prevents the re-weighting logic of the Records class from working correctly in 2014.

I'm not sure how any of the 2014 `cps.csv` tax results can be correct given this misnaming of the first column.

**Comment by hdoupe** [comment]
@martinholmer @andersonfrailey Just to clarify, we have weights for the CPS until 2027, but there is an extra variable, `SEQUENCE`.  Is this correct?

**Comment by martinholmer** [comment]
@hdoupe said:
> Just to clarify, we have weights for the CPS until 2027, but there is an extra variable, SEQUENCE. Is this correct?

No, that is not correct.  What I said was this:
> Notice the bogus variable name for the first column in the `cps_weights.csv` file.
> It should be `WT2014` not `SEQUENCE`.
> And that **renamed column** should contain the weights for 2014.

The 2014 weights are in the `s006` variable in the `cps.csv` file.  
They need to be copied to the renamed first column of the `cps_weights.csv` file.

**Comment by hdoupe** [comment]
@martinholmer Ok, this makes sense.  So `SEQUENCE` needs to be replaced by the `s006` variable and we do indeed have the weights for 2027.

**Comment by andersonfrailey** [comment]
Thanks for pointing this bug out, @martinholmer. I'll modify the weights final prep script to drop `SEQUENCE`.

**Comment by martinholmer** [comment]
@andersonfrailey said:
> I'll modify the weights final prep script to drop SEQUENCE.

Yes, but that's only part of the needed repair.  The `cps_weights.csv.gz` file need to have a column labeled `WT2014` that contains the `s006` values from the `cps.csv.gz` file.  Otherwise the sampling logic will not work for 2014.

**Comment by andersonfrailey** [comment]
@martinholmer said:
>Yes, but that's only part of the needed repair. The cps_weights.csv.gz file needs to have a column labeled WT2014 that contains the s006 values from the cps.csv.gz file. Otherwise the sampling logic will not work for 2014.

Got it. I'll fix it this afternoon and open up a PR.

**Comment by andersonfrailey** [comment]
Quick update. There likely won't be a PR fixing this today. While working on it, I realized that the s006 variable in the CPS is the same as the `WT2015` column in the weights file. After originally speaking to John, I was under the impression that, because the CPS file is representative of tax year 2014, the `s006` variable in the CPS was equivalent to a column like `WT2014`, but this seems to not be the case. My theory is, because the newest CPS used to create the file is the March 2015 supplement (representative of tax year 2014), John labeled those weights as `WT2015`, and then incremented from there.

I've emailed him to ask for clarification, but if I'm correct then each weight year would be off by one. So `WT2016` should be `WT2015`, `WT2017` should be `WT2018`, and so on. I will update this when I hear back from John.

**Comment by martinholmer** [comment]
@andersonfrailey said ten days ago:
> Quick update. There likely won't be a PR fixing this today. While working on it, I realized that the s006 variable in the CPS is the same as the WT2015 column in the weights file. After originally speaking to John, I was under the impression that, because the CPS file is representative of tax year 2014, the s006 variable in the CPS was equivalent to a column like WT2014, but this seems to not be the case. My theory is, because the newest CPS used to create the file is the March 2015 supplement (representative of tax year 2014), John labeled those weights as WT2015, and then incremented from there.
>
> I've emailed him to ask for clarification, but if I'm correct then each weight year would be off by one. So WT2016 should be WT2015, WT2017 should be WT2018, and so on. I will update this when I hear back from John.

@andersonfrailey, Have you head anything from John about this?

**Comment by andersonfrailey** [comment]
@martinholmer I have not. I'll email him again.

**Comment by martinholmer** [comment]
@andersonfrailey, 
What does the Fortran code in pull request #145 imply about the bug discussed in taxdata issue #125?

**Comment by andersonfrailey** [comment]
@martinholmer, John did say that he mislabeled the weights. He was a little vague in what exactly that implies about the bug discussed in this issue though. I've emailed him again for clarification and am going to try and schedule a phone call with him next week.

**Comment by martinholmer** [comment]
@andersonfrailey said:
> John did say that he mislabeled the [CPS] weights. He was a little vague in what exactly that implies about the bug discussed in this issue though. I've emailed him again for clarification and am going to try and schedule a phone call with him next week.

Good!  Slow progress is better than no progress.

**Comment by martinholmer** [comment]
@andersonfrailey 
What's the status of taxdata issue #125?
We have evidence that the 2014 weights in the `cps.csv.gz` file are bad.
What, if anything, are we going to do about that?

@MattHJensen, do you want to leave the 2014 CPS data unusable?
If so, you should put a notice somewhere about the 2014 weight problem.

**Comment by andersonfrailey** [comment]
On my end, I haven't been able to get in touch with John to confirm anything about the weights file. But my thought is that it's not the 2014 weighted in `cps.csv.gz` that are bad - those come directly  from the CPS files - it is the 2015 weights in `cps_weights.csv.gz` that have an issue.

**Comment by martinholmer** [comment]
@andersonfrailey said:
> On my end, I haven't been able to get in touch with John to confirm anything about the weights file. But my thought is that it's not the 2014 weighted in `cps.csv.gz` that are bad - those come directly from the CPS files - it is the 2015 weights in `cps_weights.csv.gz` that have an issue.

All I know is that the 2014 weights are such that the sub-sampling test fails for 2014 when it passes for each of the nine years beginning with 2015.  See `test_agg` in the `tax-calculator/taxcalc/tests/test_cpscsv.py` file (search for the phrase TODO to find the place where test fails).

**Comment by martinholmer** [comment]
@andersonfrailey said:
> 2014 weights in `cps.csv.gz` ... - those come directly from the CPS files

But my recollection is that you told me John said there was some confusion because of the fact that **three** CPS samples, each for a different year, are combined in the `cps.csv.gz` file.  How are the weights for the other two years adjusted to be compatible with the 2014 weights?

**Comment by martinholmer** [comment]
Issue #125 has been resolved by pull request #200.
